File: Ch05/05_03/wg.py
from queue import Queue
from threading import Thread
from time import sleep


def worker(id, queue):
    while True:
        item = queue.get()
        sleep(0.1)
        print(f'{id} finished {item}')
        queue.task_done()


if __name__ == '__main__':
    queue = Queue()
    for id in range(3):
        Thread(target=worker, args=(id, queue), daemon=True).start()

    for msg in ['A', 'B', 'C', 'D', 'E', 'F']:
        queue.put(msg)

    queue.join()
    print('all jobs done')

File: Ch05/05_03/wg.go
package main

import (
	"fmt"
	"sync"
	"time"
)

func main() {
	var wg sync.WaitGroup
	ch := make(chan string)

	for i := 0; i < 3; i++ {
		go func(id int) {
			for msg := range ch {
				time.Sleep(100 * time.Millisecond)
				fmt.Printf("%d finished %s\n", id, msg)
				wg.Done()
			}
		}(i)
	}

	for _, msg := range []string{"A", "B", "C", "D", "E", "F"} {
		wg.Add(1)
		ch <- msg
	}
	wg.Wait()
	fmt.Println("all jobs done")
}
